src/3d/qgs3dmodelcache.cpp

```diff
@@ -0,0 +1,147 @@
+/***************************************************************************
+                         qgs3dmodelcache.cpp
+                         -----------------
+    begin                : July 2020
+    copyright            : (C) 2020 by Mathieu Pellerin
+    email                : nirvn dot asia at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgs3dmodelcache.h"
+
+#include "qgis.h"
+#include "qgslogger.h"
+
+#include <QFile>
+#include <QBuffer>
+
+///@cond PRIVATE
+
+Qgs3DModelCacheEntry::Qgs3DModelCacheEntry( const QString &path )
+  : QgsAbstractContentCacheEntry( path )
+{
+}
+
+bool Qgs3DModelCacheEntry::isEqual( const QgsAbstractContentCacheEntry *other ) const
+{
+  const Qgs3DModelCacheEntry *other3dModel = dynamic_cast< const Qgs3DModelCacheEntry * >( other );
+  // cheapest checks first!
+  if ( !other3dModel || other3dModel->base64 != base64 )
+    return false;
+
+  return true;
+}
+
+int Qgs3DModelCacheEntry::dataSize() const
+{
+  return base64.size();
+}
+
+void Qgs3DModelCacheEntry::dump() const
+{
+  QgsDebugMsgLevel( QStringLiteral( "path: %1" ).arg( path ), 3 );
+}
+
+///@endcond
+
+// canonical 3D model cache instance
+Qgs3DModelCache *Qgs3DModelCache::s3dModelCache = nullptr;
+
+Qgs3DModelCache::Qgs3DModelCache( QObject *parent )
+  : QgsAbstractContentCache< Qgs3DModelCacheEntry >( parent, QObject::tr( "3D Model" ) )
+{
+  connect( this, &QgsAbstractContentCacheBase::remoteContentFetched, this, &Qgs3DModelCache::remote3DModelFetched );
+}
+
+Qt3DRender::QMesh *Qgs3DModelCache::pathAsMesh( const QString &path, bool blocking )
+{
+  Qt3DRender::QMesh *mesh = new Qt3DRender::QMesh;
+
+  const QString file = path.trimmed();
+  if ( file.isEmpty() )
+    return mesh;
```

Comment:
should this be nullptr?

Reply:
Na, it returns an empty mesh object. It's IMHO better than nullptr, it gives you the certainty that you get an object. You can always is the source().isEmpty() afterwards.